machinist/mserver/mserver.go

```diff
@@ -45,12 +47,14 @@ func (s *ControlPlane) Run() error {
 		s.killChannel <- s.Controller.dnsServer.Run()
 	}()
 	s.Controller.Init()
-	go s.Controller.ServeAllRecords()
+	go s.Controller.ServeAllRecords(s.allRecordsKillChannel)
 	go s.Controller.WriteState()
 	return grpcs.Serve(s.Listener)
 }
 
 func (s *ControlPlane) Stop() error {
+	s.allRecordsKillChannel <- struct{}{}
+	<-s.allRecordsKillChannel
```

Comment:
I think this should be two separate channels unfortunately; I don't think there's anything that stops the receive from picking up the kill signal inserted on the previous line.